### DISCUSSION TITLE
ETSI DENM Message

### DISCUSSION THREAD

**Discussion by iTacco** [eclipse/mosaic#56] [Q&A] [answered]
Hello everybody,
I have a question about the etsi denm messsage send by the "V2xMessageTransmission" interaction inside the Barnim scenario.

This is how the whole V2xMessageTransmission interaction looks like in json:
`{"V2xMessageTransmission":{"message":{"payload":{"contentClassName":"org.eclipse.mosaic.lib.objects.v2x.etsi.DenmContent","bytes":[0,0,0,-25,-26,57,-24,0,64,74,83,60,12,31,-56,-13,64,43,23,10,61,112,-93,-41,0,0,0,0,0,0,0,0,1,0,27,45,51,51,54,54,95,50,48,50,54,51,54,50,57,52,48,95,49,51,49,51,56,56,53,53,48,50,1,0,64,-34,56,-28,0,0,0,0,1,64,74,81,7,-81,32,-22,91,64,43,33,112,-42,43,-15,32,0,0,0,0,0,0,0,0,0,0,0,0,0],"lengthInBytes":102,"minimalLength":200},"denmContent":{"time":996000000000,"senderPosition":{"latitude":52.65027,"longitude":13.545},"eventRoadId":"-3366_2026362940_1313885502","warningType":"ICE","eventStrength":0,"causedSpeed":6.9444447,"senderDeceleration":0.0,"eventLocation":{"latitude":52.633047,"longitude":13.565314}},"routing":{"source":{"sourceAddress":{"address":"10.2.0.0"},"sourceName":"rsu_0","sourcePosition":{"latitude":52.65027,"longitude":13.545}},"destination":{"destinationAddress":{"address":"255.255.255.255"},"type":"CELL_GEOCAST","disseminationArea":{"center":{"latitude":52.633047,"longitude":13.565314},"radius":3000.0},"protocolType":"UDP"}},"id":3141,"sequenceNumber":499},"time":996000000000,"id":52172,"senderId":"application","typeId":"V2xMessageTransmission"}}`

But I am interested in the "bytes" array: `"bytes":[0,0,0,-25,-26,57,-24,0,64,74,83,60,12,31,-56,-13,64,43,23,10,61,112,-93,-41,0,0,0,0,0,0,0,0,1,0,27,45,51,51,54,54,95,50,48,50,54,51,54,50,57,52,48,95,49,51,49,51,56,56,53,53,48,50,1,0,64,-34,56,-28,0,0,0,0,1,64,74,81,7,-81,32,-22,91,64,43,33,112,-42,43,-15,32,0,0,0,0,0,0,0,0,0,0,0,0,0]`.

As I am expecting this byte array should represent a DENM byte message in the "ETSI DENM EN 302 637-3 v1.3.1" format.
The array is a list of signed integers which cannot be decoded by following online decoder: https://www.marben-products.com/decoder-asn1-automotive/
So I transformed the array in something similar to this:
`Decimal stream:

0, 0, 0, -24, 93, 111, 124, 0, 64, 74, 83, 60, 12, 31, -56, -13, 64, 43, 23, 10, 61, 112, -93, -41, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 27, 45, 51, 51, 54, 54, 95, 50, 48, 50, 54, 51, 54, 50, 57, 52, 48, 95, 49, 51, 49, 51, 56, 56, 53, 53, 48, 50, 1, 0, 64, -34, 56, -28, 0, 0, 0, 0, 1, 64, 74, 81, 7, -81, 32, -22, 91, 64, 43, 33, 112, -42, 43, -15, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0

to this hex stream:
00 00 00 e8 5d 6f 7c 00 40 4a 53 3c 0c 1f c8 f3 40 2b 17 0a 3d 70 a3 29 00 00 00 00 00 00 00 00 01 00 1b 2d 33 33 36 36 5f 32 30 32 36 33 36 32 39 34 30 5f 31 33 31 33 38 38 35 35 30 32 01 00 40 de 38 e4 00 00 00 00 01 40 4a 51 07 af 20 ea 5b 40 2b 21 70 d6 2b f1 20 00 00 00 00 00 00 00 00 00 00 00 00 00`

and put it into the online decoder shown above. But the output isn't a valid DENM message.

So I am asking myself if I had done something wrong or this isn't a valid ETSI DENM message at all or I have to handle the byte array in a different way.

Thanks in advance,
iTacco

**Comment by rprotzmann** [comment, marked as answer]
Regarding the modelling of ETSI standardized V2X messages like DENM and CAM, we focused primarily on the content.

Our main goals are:
-	Incorporation of most important fields and containers of the messages
-	Flexibility to include additional content, which is not standardized up to now (`extendedContainers` for DENM, `userTaggedValue` for CAM)
-	Support of advanced sending schemes with (periodic) retransmissions
-	Performance for simulation of many vehicles / nodes and messages

Due to the last goal, we currently do _NOT_ employ ASN.1 formatting in the provided message handling.

Moreover, we usually cache the message payload in the Application Simulator and in the `V2xMessageTransmission` only hand over Meta information for the message like `Id`, `Routing`, `Timestamp` and `Length`, which are needed for Communication Simulation. On reception, the Application Simulator fetches the payload from its cache and distributes it to the receiving entities.

The `EncodedPayload` allows for publishing of the payload as serialized Java object in the `V2xMessageTransmission`, if needed in a simulator or logger. Yet, it is not ASN.1 formatted.
